### PR TITLE
Stream cantabular CSV and add a checker func for Cantabular API Ext

### PR DESCRIPTION
**Pull request by DavidSubiros**
### What

Stream CSV response from Cantabular so that we don't need to keep the whole response in memory.

Trello card: https://trello.com/c/Ph3rIO3M/5354-update-dp-cantabular-csv-exporter-to-use-latest-version-of-code-from-scc

SCC code reference
https://github.com/cantabular/examples/tree/master/golang/cmd/cantabular-query-streamed

- Implemented generic Stream function that uses an io.Pipe to read, transform, and write on the fly (used @redhug1  Cantabular XLSX implementation as a reference)
- Copied JSON streamer from SCC, added error handling, and supported a few extra corner cases for 'data' being decoded before 'error'
- Implemented a new function for Cantabular client, called `StaticDatasetQueryStreamCSV`, where the caller only needs to provide the Consumer method that will consume the CSV stream.
- Added unit tests

Required by [dp-cantabular-csv-exporter](https://github.com/ONSdigital/dp-cantabular-csv-exporter/pull/42)

### How to review

- Make sure code changes make sense
- Make sure the new method `StaticDatasetQueryStreamCSV` is easy to use by a caller and gives enough flexibility
- Make sure unit test pass
- (info only, not required - test with a real client):
  - You can start Cantabular server and api ext docker conainers
  - You can replace the unit-test mock with a real http client, and the unit tests should work as expected:
```
		cantabularClient := cantabular.NewClient(
			cantabular.Config{
				Host:           "http://localhost:8491",
				ExtApiHost:     "http://localhost:8492",
				GraphQLTimeout: 10 * time.Second,
			},
			dphttp.NewClient(),
			nil,
		)
```

### Who can review

anyone

### CODE REVIEW

File: cantabular/table.go [hunk 1]
```diff
@@ -54,22 +64,243 @@ func (c *Client) ParseTable(table Table) (*bufio.Reader, error) {
...
+
+	// find starting '{'
+	isStartObj, err := dec.StartObjectComposite()
+	if err != nil {
+		return fmt.Errorf("error decoding start of json object: %w", err)
+	}
+	if !isStartObj {
+		return errors.New("no json object found in response")
+	}
+
```
Comment (ChrysmOre) [inline cantabular/table.go:83]: had an experiment, maybe the second version looks a little neater?
```
isStartObj, err := dec.StartObjectComposite()
if err != nil {
	return fmt.Errorf("error decoding start of json object: %w", err)
}
if !isStartObj {
	return errors.New("no json object found in response")
}
```
or
```
if isStartObj, err := dec.StartObjectComposite(); err != nil {
	return fmt.Errorf("error decoding start of json object: %w", err)
} else if !isStartObj {
	return errors.New("no json object found in response")
}
```

File: cantabular/client.go [hunk 2]
```diff
@@ -69,22 +71,63 @@ func (c *Client) httpGet(ctx context.Context, path string) (*http.Response, erro
 		return nil, dperrors.New(
 			fmt.Errorf("failed to make request: %w", err),
 			http.StatusInternalServerError,
+			log.Data{
+				"url":    path,
+				"method": "get",
```
[the remark this answers is not in the document]
Comment (DavidSubiros) [reply cantabular/client.go:76]: according to the standards, the event data should be lower case: https://github.com/ONSdigital/dp/blob/main/standards/LOGGING_STANDARDS.md#effective-event-types
But they allow some expections, I think this one is an exception: https://github.com/ONSdigital/dp/blob/main/standards/LOGGING_STANDARDS.md#http-event-data
I will change it

File: cantabular/static_dataset.go [hunk 3]
```diff
@@ -1,25 +1,49 @@
...
+
 // StaticDataset represents the 'dataset' field from a GraphQL static dataset
 // query response
 type StaticDataset struct {
 	Table Table `json:"table" graphql:"table(variables: $variables)"`
 }
 
+// graphQLQuery represents a full query to be encoded into the body of a plain http post request
+// for a GraphQL static dataset query
+const graphQLQuery = `
```
[the remark this answers is not in the document]
Comment (DavidSubiros) [reply cantabular/static_dataset.go:29]: maybe `QueryStaticDataset`?

File: cantabular/table.go (hunk 1 above)
Comment (DavidSubiros) [reply cantabular/table.go:83]: I agree, I will replace it where needed.

**Review by ChrysmOre — APPROVED** [review]

**Pull request merged**